src/dune_rules/melange_rules.ml

```diff
@@ -1,6 +1,37 @@
 open Import
 module CC = Compilation_context
 
+let melc sctx ~dir =
+  (* TODO loc should come from the mode field in the dune file *)
+  Super_context.resolve_program sctx ~loc:None ~dir ~hint:"opam install melange"
+    "melc"
+
+let melange_where =
+  let impl bin =
+    let open Memo.O in
+    let* _ = Build_system.build_file bin in
+    let+ where =
+      Memo.of_reproducible_fiber
+      @@ Process.run_capture_line Process.Strict bin [ "--where" ]
+    in
+    Path.of_string where
+  in
+  let memo =
+    Memo.create "melange-where" ~input:(module Path) ~cutoff:Path.equal impl
+  in
+  fun sctx ~dir ->
+    let open Memo.O in
+    let* env = Super_context.env_node sctx ~dir >>= Env_node.external_env in
+    match Env.get env "MELANGELIB" with
+    | Some p -> Memo.return (Some (Path.of_string p))
```

Comment:
for the record, @emillon and I ran into some cases where `MELANGELIB` would be set to empty string. Not sure if it should be handled here. Or it is enough with https://github.com/melange-re/melange/issues/412.